File: 2024/day12/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <set>
#include <queue>
#include <utility>

using namespace std;

struct Grid {
    vector<string> data;
    int rows;
    int cols;

    Grid(const vector<string>& lines) : data(lines) {
        rows = data.size();
        cols = data.empty() ? 0 : data[0].size();
    }

    char at(int r, int c) const {
        if (r < 0 || r >= rows || c < 0 || c >= cols) return '\0';
        return data[r][c];
    }
};

using Cell = pair<int, int>;
using Region = set<Cell>;

vector<Region> find_regions(const Grid& grid) {
    set<Cell> visited;
    vector<Region> regions;

    const int dr[] = {0, 0, 1, -1};
    const int dc[] = {1, -1, 0, 0};

    for (int r = 0; r < grid.rows; r++) {
        for (int c = 0; c < grid.cols; c++) {
            if (visited.count({r, c})) continue;

            // BFS to find all cells in this region
            char plant = grid.at(r, c);
            Region region;
            queue<Cell> q;
            q.push({r, c});

            while (!q.empty()) {
                auto [cr, cc] = q.front();
                q.pop();

                if (visited.count({cr, cc})) continue;
                if (cr < 0 || cr >= grid.rows || cc < 0 || cc >= grid.cols) continue;
                if (grid.at(cr, cc) != plant) continue;

                visited.insert({cr, cc});
                region.insert({cr, cc});

                for (int i = 0; i < 4; i++) {
                    int nr = cr + dr[i];
                    int nc = cc + dc[i];
                    if (!visited.count({nr, nc})) {
                        q.push({nr, nc});
                    }
                }
            }

            regions.push_back(region);
        }
    }

    return regions;
}

int calculate_perimeter(const Region& region) {
    int perimeter = 0;
    const int dr[] = {0, 0, 1, -1};
    const int dc[] = {1, -1, 0, 0};

    for (const auto& [r, c] : region) {
        for (int i = 0; i < 4; i++) {
            int nr = r + dr[i];
            int nc = c + dc[i];
            if (!region.count({nr, nc})) {
                perimeter++;
            }
        }
    }

    return perimeter;
}

int count_sides(const Region& region) {
    int corners = 0;

    for (const auto& [r, c] : region) {
        // Check all 8 neighbors
        bool up = region.count({r - 1, c});
        bool down = region.count({r + 1, c});
        bool left = region.count({r, c - 1});
        bool right = region.count({r, c + 1});
        bool up_left = region.count({r - 1, c - 1});
        bool up_right = region.count({r - 1, c + 1});
        bool down_left = region.count({r + 1, c - 1});
        bool down_right = region.count({r + 1, c + 1});

        // Top-left corner
        if (!up && !left) corners++;  // convex
        else if (up && left && !up_left) corners++;  // concave

        // Top-right corner
        if (!up && !right) corners++;  // convex
        else if (up && right && !up_right) corners++;  // concave

        // Bottom-left corner
        if (!down && !left) corners++;  // convex
        else if (down && left && !down_left) corners++;  // concave

        // Bottom-right corner
        if (!down && !right) corners++;  // convex
        else if (down && right && !down_right) corners++;  // concave
    }

    return corners;
}

long long part1(const Grid& grid) {
    auto regions = find_regions(grid);
    long long total = 0;

    for (const auto& region : regions) {
        int area = region.size();
        int perimeter = calculate_perimeter(region);
        total += static_cast<long long>(area) * perimeter;
    }

    return total;
}

long long part2(const Grid& grid) {
    auto regions = find_regions(grid);
    long long total = 0;

    for (const auto& region : regions) {
        int area = region.size();
        int sides = count_sides(region);
        total += static_cast<long long>(area) * sides;
    }

    return total;
}

int main() {
    // Read input
    ifstream infile("../input.txt");
    if (!infile) {
        cerr << "Error opening input file" << endl;
        return 1;
    }

    vector<string> lines;
    string line;
    while (getline(infile, line)) {
        if (!line.empty()) {
            lines.push_back(line);
        }
    }

    Grid grid(lines);

    cout << "Part 1: " << part1(grid) << endl;
    cout << "Part 2: " << part2(grid) << endl;

    return 0;
}

File: 2024/day12/rust/solution.rs
use std::collections::{HashSet, VecDeque};
use std::fs;

fn main() {
    let input = fs::read_to_string("../input.txt")
        .expect("Failed to read input file");

    let grid: Vec<Vec<char>> = input
        .trim()
        .lines()
        .map(|line| line.chars().collect())
        .collect();

    let rows = grid.len();
    let cols = grid[0].len();

    println!("Part 1: {}", part1(&grid, rows, cols));
    println!("Part 2: {}", part2(&grid, rows, cols));
}

fn find_regions(grid: &Vec<Vec<char>>, rows: usize, cols: usize) -> Vec<HashSet<(i32, i32)>> {
    let mut visited = HashSet::new();
    let mut regions = Vec::new();

    for r in 0..rows {
        for c in 0..cols {
            let pos = (r as i32, c as i32);
            if visited.contains(&pos) {
                continue;
            }

            // BFS to find all cells in this region
            let plant = grid[r][c];
            let mut region = HashSet::new();
            let mut queue = VecDeque::new();
            queue.push_back(pos);

            while let Some((cr, cc)) = queue.pop_front() {
                if visited.contains(&(cr, cc)) {
                    continue;
                }
                if cr < 0 || cr >= rows as i32 || cc < 0 || cc >= cols as i32 {
                    continue;
                }
                if grid[cr as usize][cc as usize] != plant {
                    continue;
                }

                visited.insert((cr, cc));
                region.insert((cr, cc));

                for (dr, dc) in [(0, 1), (0, -1), (1, 0), (-1, 0)] {
                    let nr = cr + dr;
                    let nc = cc + dc;
                    if !visited.contains(&(nr, nc)) {
                        queue.push_back((nr, nc));
                    }
                }
            }

            regions.push(region);
        }
    }

    regions
}

fn calculate_perimeter(region: &HashSet<(i32, i32)>) -> usize {
    let mut perimeter = 0;

    for (r, c) in region {
        for (dr, dc) in [(0, 1), (0, -1), (1, 0), (-1, 0)] {
            let nr = r + dr;
            let nc = c + dc;
            if !region.contains(&(nr, nc)) {
                perimeter += 1;
            }
        }
    }

    perimeter
}

fn count_sides(region: &HashSet<(i32, i32)>) -> usize {
    let mut corners = 0;

    for (r, c) in region {
        // Check all 4 corners of this cell
        // Each corner is defined by checking two orthogonal neighbors and the diagonal
        // Convex: both orthogonal out
        // Concave: both orthogonal in, diagonal out

        let up = region.contains(&(r - 1, *c));
        let down = region.contains(&(r + 1, *c));
        let left = region.contains(&(*r, c - 1));
        let right = region.contains(&(*r, c + 1));
        let up_left = region.contains(&(r - 1, c - 1));
        let up_right = region.contains(&(r - 1, c + 1));
        let down_left = region.contains(&(r + 1, c - 1));
        let down_right = region.contains(&(r + 1, c + 1));

        // Top-left corner
        if !up && !left {
            corners += 1; // convex
        } else if up && left && !up_left {
            corners += 1; // concave
        }

        // Top-right corner
        if !up && !right {
            corners += 1; // convex
        } else if up && right && !up_right {
            corners += 1; // concave
        }

        // Bottom-left corner
        if !down && !left {
            corners += 1; // convex
        } else if down && left && !down_left {
            corners += 1; // concave
        }

        // Bottom-right corner
        if !down && !right {
            corners += 1; // convex
        } else if down && right && !down_right {
            corners += 1; // concave
        }
    }

    corners
}

fn part1(grid: &Vec<Vec<char>>, rows: usize, cols: usize) -> usize {
    let regions = find_regions(grid, rows, cols);
    let mut total = 0;

    for region in regions {
        let area = region.len();
        let perimeter = calculate_perimeter(&region);
        total += area * perimeter;
    }

    total
}

fn part2(grid: &Vec<Vec<char>>, rows: usize, cols: usize) -> usize {
    let regions = find_regions(grid, rows, cols);
    let mut total = 0;

    for region in regions {
        let area = region.len();
        let sides = count_sides(&region);
        total += area * sides;
    }

    total
}
